### PR TITLE
MAM refactoring - part 2

### PR DESCRIPTION
**Pull request by michalwski**
Proposed changes include:
* unify message forwarding for different MAM versions
* move `pm` and `muc` duplicated helper function to `mam_iq`
* add possibility to configure a callback module which can add extra lookup params based on custom form fields

### CODE REVIEW

File: apps/ejabberd/src/mod_mam.erl [hunk 1]
```diff
@@ -244,10 +243,10 @@ stop(Host) ->
 -spec process_mam_iq(From :: ejabberd:jid(), To :: ejabberd:jid(),
                      IQ :: ejabberd:iq()) -> ejabberd:iq() | ignore.
 process_mam_iq(From=#jid{lserver=Host}, To, IQ) ->
-    Action = iq_action(IQ),
+    Action = mam_iq:action(IQ),
     case is_action_allowed(Action, From, To) of
         true  ->
-            case wait_shaper(Host, Action, From) of
+            case mam_iq:wait_shaper(Host, Action, From) of
```
[the remark this answers is not in the document]
Comment (michalwski) [reply apps/ejabberd/src/mod_mam.erl]: That's right. I'll move it to mod_mam_utils.

File: apps/ejabberd/src/mod_mam_meta.erl [hunk 2]
```diff
@@ -88,10 +88,11 @@ valid_core_mod_opts(mod_mam) ->
     [add_archived_element,
      is_archivable_message,
      archive_chat_markers,
+     extra_lookup_params,
      full_text_search,
      archive_groupchats];
 valid_core_mod_opts(mod_mam_muc) ->
-    [add_archived_element, is_archivable_message, host, full_text_search].
+    [add_archived_element, is_archivable_message, host, extra_lookup_params, full_text_search].
```
[the remark this answers is not in the document]
Comment (michalwski) [reply apps/ejabberd/src/mod_mam_meta.erl]: Correct. I'll try to do it as soon as possible.

**Comment by michalwski** [comment]
I updated the PR with changes requested in review.

File: apps/ejabberd/src/mod_mam_meta.erl (hunk 2 above)
[the remark this answers is not in the document]
Comment (michalwski) [reply apps/ejabberd/src/mod_mam_meta.erl]: The doc was added.

File: doc/modules/mod_mam.md [hunk 3]
```diff
@@ -31,6 +31,10 @@ For now `odbc` backend has very limited support for this feature, while `cassand
 * **archive_chat_markers** (boolean, default: `false`) - If set to true, XEP-0333 chat markers will be archived. See more details [here](#archiving-chat-markers)
 * **pm** (list | `false`, default: `[]`) - Override options for archivization of one-to-one messages. If the value of this option is `false`, one-to-one message archive is disabled.
 * **muc** (list | `false`, default: `false`) - Override options for archivization of group chat messages. If the value of this option is `false`, group chat message archive is disabled.
+* **extra_lookup_params** (atom, default: `undefined`) - a module implementing `mam_iq` behaviour.
+ If this option has value other then undefined, function `extra_lookup_params/2` from this module will be called when building MAM lookup parameters.
```
Comment (goddammit) [inline doc/modules/mod_mam.md]: When building MAM lookup parameters the function `extra_lookup_params/2` is called  for values other than 'undefined'?

File: doc/modules/mod_mam.md [hunk 4]
```diff
@@ -31,6 +31,10 @@ For now `odbc` backend has very limited support for this feature, while `cassand
...
+ This can be used to extend currently supported MAM query fields by some custom field.
```
Comment (goddammit) [inline doc/modules/mod_mam.md]: This can be used to extend currently supported MAM query fields by a custom field.

File: doc/modules/mod_mam.md [hunk 5]
```diff
@@ -31,6 +31,10 @@ For now `odbc` backend has very limited support for this feature, while `cassand
...
+ This field can be added to lookup params which are later passed to MAM backend.
```
Comment (goddammit) [inline doc/modules/mod_mam.md]: This field can be added to lookup params later passed to the MAM backend.

File: doc/modules/mod_mam.md [hunk 6]
```diff
@@ -70,9 +74,9 @@ These options will only have effect when the `odbc` backend is used:
 #### Common backend options
 
 * **user_prefs_store** (atom, default: `false`) - Leaving this option as `false` will prevent users from setting their archiving preferences. It will also increase performance. Other possible values are:
-    * `odbc` (ODBC backend only) - User archiving preferences saved in ODBC. Slow and not recommended, but might be used to simplify things and keep everything in ODBC.
-    * `cassandra` (Cassandra backend only) - User archiving preferences are saved in Cassandra.
-    * `mnesia` (recommended) - User archiving preferences saved in Mnesia and accessed without transactions. Recommended in most deployments, could be overloaded with lots of users updating their preferences at once. There's a small risk of an inconsistent (in a rather harmless way) state of the preferences table.
+  * `odbc` (ODBC backend only) - User archiving preferences saved in ODBC. Slow and not recommended, but might be used to simplify things and keep everything in ODBC.
```
Comment (goddammit) [inline doc/modules/mod_mam.md]: User archiving preferences saved in ODBC. Slow and not recommended, but might be used for simplicity (keeping everything in ODBC).
?

File: doc/modules/mod_mam.md (hunk 3 above)
Comment (michalwski) [reply doc/modules/mod_mam.md]: It's not clear to me in the way you suggested @goddammit. What I wanted to say is exactly the following.

If the value of `extra_lookup_params` is other then `undefined`, it means this is a `module`. The function `extra_lookup_params/2` from the `module` will be called when building MAM lookup parameters.

Hope this explains my intention and helps to build better sentence if needed :)

**Pull request merged**